internal/dbt/service.go

```diff
@@ -0,0 +1,413 @@
+	}))
+
+	grantIds := make(map[string]string)
+	maskIds := make(map[string]string)
+	filterIds := make(map[string]string)
+	apsToRemove := set.NewSet[string]()
+
+	for existingAp := range existingAps {
+		if existingAp.HasError() {
+			return nil, nil, nil, nil, fmt.Errorf("list access providers: %w", existingAp.GetError())
+		}
+
+		ap := existingAp.GetItem()
+		switch ap.Action {
+		case models.AccessProviderActionGrant:
+			if _, found := grants[ap.Name]; found {
+				if _, idFound := grantIds[ap.Name]; idFound {
+					apsToRemove.Add(ap.Id) // Remove ap with same name
+				} else {
+					grantIds[ap.Name] = ap.Id
+				}
+			} else {
+				apsToRemove.Add(ap.Id)
+			}
+		case models.AccessProviderActionFiltered:
+			if _, found := filters[ap.Name]; found {
+				if _, idFound := filterIds[ap.Name]; idFound {
+					apsToRemove.Add(ap.Id) // Remove ap with same name
+				} else {
+					filterIds[ap.Name] = ap.Id
+				}
+			} else {
+				apsToRemove.Add(ap.Id)
+			}
+		case models.AccessProviderActionMask:
+			if _, found := masks[ap.Name]; found {
+				if _, idFound := maskIds[ap.Name]; idFound {
+					apsToRemove.Add(ap.Id) // Remove ap with same name
+				} else {
+					maskIds[ap.Name] = ap.Id
+				}
+			} else {
+				apsToRemove.Add(ap.Id)
+			}
+		default:
+			continue
+		}
+	}
+
+	return grantIds, filterIds, maskIds, apsToRemove, nil
+}
+
+func (s *DbtService) loadDbtFile(dbtFilePath string) (*types.Manifest, error) {
+	jsonBytes, err := os.ReadFile(dbtFilePath)
+	if err != nil {
+		return nil, fmt.Errorf("reading dbt file: %w", err)
+	}
+
+	var result types.Manifest
+
+	err = json.Unmarshal(jsonBytes, &result)
+	if err != nil {
+		return nil, fmt.Errorf("parsing dbt file: %w", err)
+	}
+
+	return &result, nil
+}
+
+func (s *DbtService) loadAccessProvidersFromManifest(manifest *types.Manifest) (map[string]*sdkTypes.AccessProviderInput, map[string]*sdkTypes.AccessProviderInput, map[string]*sdkTypes.AccessProviderInput, error) {
+	source := _source(manifest.Metadata.ProjectName)
+
+	grants := make(map[string]*sdkTypes.AccessProviderInput)
+	filters := make(map[string]*sdkTypes.AccessProviderInput)
+	masks := make(map[string]*sdkTypes.AccessProviderInput)
+
+	var err error
+
+	defaultLocks := []sdkTypes.AccessProviderLockDataInput{
+		{
+			LockKey: sdkTypes.AccessProviderLockWhatlock,
+			Details: &sdkTypes.AccessProviderLockDetailsInput{
+				Reason: utils.Ptr(lockReason),
+			},
+		},
+		{
+			LockKey: sdkTypes.AccessProviderLockNamelock,
+			Details: &sdkTypes.AccessProviderLockDetailsInput{
+				Reason: utils.Ptr(lockReason),
+			},
+		},
+	}
+
+	supportedResourceTypes := set.NewSet("model", "seed", "snapshot")
+
+	for i := range manifest.Nodes {
+		if !supportedResourceTypes.Contains(manifest.Nodes[i].ResourceType) {
+			continue
+		}
+
+		databaseName := manifest.Nodes[i].Database
+		schemaName := manifest.Nodes[i].Schema
+		modelName := manifest.Nodes[i].Name
+		doName := fmt.Sprintf("%s.%s.%s", databaseName, schemaName, modelName)
```

Comment:
Just as a note: this will only work for specific data sources (this format of fullname)

Reply:
Yes you are right. For DBT I think it is a valid assumption. As all DBT integrations are databases the intented full name is always like this right?